File: python-usfm-parser/src/usfm_grammar/usj_generator.py
'''Logics for syntax-tree to dict(USJ) conversions'''
from usfm_grammar.usx_generator import USXGenerator

class USJGenerator:
    '''A binding for all methods used in generating USJ from Syntax tree'''

    PARA_STYLE_MARKERS = USXGenerator.PARA_STYLE_MARKERS
    NOTE_MARKERS = USXGenerator.NOTE_MARKERS
    CHAR_STYLE_MARKERS = USXGenerator.CHAR_STYLE_MARKERS
    NESTED_CHAR_STYLE_MARKERS = USXGenerator.NESTED_CHAR_STYLE_MARKERS
    DEFAULT_ATTRIB_MAP = USXGenerator.DEFAULT_ATTRIB_MAP
    TABLE_CELL_MARKERS = USXGenerator.TABLE_CELL_MARKERS
    MISC_MARKERS = USXGenerator.MISC_MARKERS

    def __init__(self, tree_sitter_language_obj, usj_root_obj=None):
        '''Initialzes the USJ generator with USFM and root object'''
        self.usfm_language = tree_sitter_language_obj
        if usj_root_obj is None:
            self.json_root_obj = {
                "type": "USJ",
                "version": "0.0.1-alpha.2",
                "content":[]
            }
        else:
            self.json_root_obj = usj_root_obj

    def findlast_from_json(self, json_obj, type_value):
        '''Traverse the given JSON and list all elements with given value in type field'''
        output = None
        if json_obj['type'] == type_value or type_value in json_obj['type'].split(':'):
            output = json_obj
        for child in json_obj['content']:
            if isinstance(child, str):
                continue
            child_output = self.findlast_from_json(child, type_value)
            if child_output is not None:
                output = child_output
        return output

    def node_2_usj_id(self, node, usfm_bytes, parent_json_obj):
        '''build id node in USX'''
        id_captures = self.usfm_language.query('''(id (bookcode) @book-code
                                                    (description)? @desc)''').captures(node)
        code = None
        desc = None
        for tupl in id_captures:
            if tupl[1] == "book-code":
                code = usfm_bytes[tupl[0].start_byte:tupl[0].end_byte].decode('utf-8')
            elif tupl[1] == 'desc':
                desc = usfm_bytes[tupl[0].start_byte:tupl[0].end_byte].decode('utf-8')
        book_json_obj = {"type": "book:id", "content":[]}
        book_json_obj['code'] = code
        if desc is not None and desc.strip() != "":
            book_json_obj['content'].append(desc.strip())
        parent_json_obj['content'].append(book_json_obj)

    def node_2_usj_c(self, node, usfm_bytes, parent_json_obj):
        '''Build c, the chapter milestone node in usj'''
        chap_cap = self.usfm_language.query('''(c (chapterNumber) @chap-num
                                            (ca (chapterNumber) @alt-num)?
                                            (cp (text) @pub-num)?)
                                        ''').captures(node)
        chap_num = usfm_bytes[chap_cap[0][0].start_byte:chap_cap[0][0].end_byte].decode('utf-8')
        chap_ref = None 
        for child in self.json_root_obj['content']:
            if child['type'] == "book:id":
                chap_ref = child['code']+" "+chap_num
                break
        chap_json_obj = {"type":"chapter:c", "content":[]}
        chap_json_obj["number"] = chap_num
        chap_json_obj["sid"] = chap_ref
        for tupl in chap_cap:
            if tupl[1] == "alt-num":
                chap_json_obj['altnumber'] =\
                    usfm_bytes[tupl[0].start_byte:tupl[0].end_byte].decode('utf-8').strip()
            if tupl[1] == "pub-num":
                chap_json_obj['pubnumber'] =\
                    usfm_bytes[tupl[0].start_byte:tupl[0].end_byte].decode('utf-8').strip()
        parent_json_obj['content'].append(chap_json_obj)

        for child in node.children:
            if child.type in ["cl", "cd"]:
                self.node_2_usj(child, usfm_bytes,parent_json_obj)

    def node_2_usj_chapter(self, node, usfm_bytes, parent_json_obj):
        '''build chapter node in USX'''
        for child in node.children:
            if child.type == "c":
                self.node_2_usj_c(child, usfm_bytes,parent_json_obj)
            else:
                self.node_2_usj(child, usfm_bytes, parent_json_obj)

    def node_2_usj_verse(self, node, usfm_bytes, parent_json_obj):
        '''build verse node in USX'''
        verse_num_cap = self.usfm_language.query('''
                                (v
                                    (verseNumber) @vnum
                                    (va (verseNumber) @alt)?
                                    (vp (text) @vp)?
                                )''').captures(node)
        verse_num = usfm_bytes[verse_num_cap[0][0].start_byte:
            verse_num_cap[0][0].end_byte].decode('utf-8')
        v_json_obj = {"type":"verse:v", "content":[]}
        for tupl in verse_num_cap:
            if tupl[1] == 'alt':
                alt_num = usfm_bytes[tupl[0].start_byte:tupl[0].end_byte].decode('utf-8')
                v_json_obj['altnumber'] = alt_num
            elif tupl[1] == 'vp':
                vp_text = usfm_bytes[tupl[0].start_byte:tupl[0].end_byte].decode('utf-8')
                v_json_obj['pubnumber'] = vp_text.strip()
        ref = str(self.findlast_from_json(self.json_root_obj, "chapter")['sid'])+ ":"+ verse_num
        v_json_obj['number'] = verse_num.strip()
        v_json_obj['sid'] = ref.strip()
        parent_json_obj['content'].append(v_json_obj)

    def node_2_usj_ca_va(self, node, usfm_bytes, parent_json_obj):
        '''Build elements for independant ca and va away from c and v'''
        style = node.type
        char_json_obj = {"type":f"char:{style}", "content":[]}
        alt_num_match = self.usfm_language.query('''([
                                            (chapterNumber)
                                            (verseNumber)
                                            ] @alt-num)''').captures(node)[0]
        alt_num = usfm_bytes[alt_num_match[0].start_byte:alt_num_match[0].end_byte] \
            .decode('utf-8').strip()
        char_json_obj["altnumber"] = alt_num
        parent_json_obj['content'].append(char_json_obj)

    def node_2_usj_para(self, node, usfm_bytes, parent_json_obj):
        '''build paragraph nodes in USX'''
        if node.children[0].type.endswith('Block'):
            for child in node.children[0].children:
                self.node_2_usj_para(child, usfm_bytes, parent_json_obj)
        elif node.type == 'paragraph':
            para_tag_cap = self.usfm_language.query(
                "(paragraph (_) @para-marker)").captures(node)[0]
            para_marker = para_tag_cap[0].type
            if not para_marker.endswith("Block"):
                para_json_obj = {"type": f"para:{para_marker}", "content":[]}
                for child in para_tag_cap[0].children[1:]:
                    self.node_2_usj(child, usfm_bytes, para_json_obj)
                parent_json_obj['content'].append(para_json_obj)
        elif node.type in ['pi', "ph"]:
            para_marker = usfm_bytes[node.children[0].start_byte:\
                node.children[0].end_byte].decode('utf-8').replace("\\", "").strip()
            para_json_obj = {"type": f"para:{para_marker}", "content":[]}
            for child in node.children[1:]:
                self.node_2_usj(child, usfm_bytes, para_json_obj)
            parent_json_obj['content'].append(para_json_obj)

    def node_2_usj_notes(self, node, usfm_bytes, parent_json_obj):
        '''build USX nodes for footnotes and corss-refs'''
        tag_node = node.children[0]
        caller_node = node.children[1]
        style = usfm_bytes[tag_node.start_byte:tag_node.end_byte].decode(
            'utf-8').replace("\\","").strip()
        note_json_obj = {"type": f"note:{style}", "content":[]}
        note_json_obj["caller"] = \
            usfm_bytes[caller_node.start_byte:caller_node.end_byte].decode('utf-8').strip()
        for child in node.children[2:-1]:
            self.node_2_usj(child, usfm_bytes, note_json_obj)
        parent_json_obj['content'].append(note_json_obj)

    def node_2_usj_char(self, node, usfm_bytes, parent_json_obj):
        '''build USX nodes for character markups, both regular and nested'''
        tag_node = node.children[0]
        # closing_node = None
        children_range = len(node.children)
        if node.children[-1].type.startswith('\\'):
            # closing_node = node.children[-1]
            children_range = children_range-1
        style = usfm_bytes[tag_node.start_byte:tag_node.end_byte].decode(
            'utf-8').replace("\\","").replace("+","").strip()
        char_json_obj = {"type": f"char:{style}", "content":[]}
        # if closing_node is None:
        #     char_json_obj["closed"] = false
        # else:
        #     char_json_obj["closed"] = true
        for child in node.children[1:children_range]:
            self.node_2_usj(child, usfm_bytes, char_json_obj)
        parent_json_obj['content'].append(char_json_obj)

    def node_2_usj_attrib(self, node, usfm_bytes, parent_json_obj):
        '''add attribute values to USX elements'''
        attrib_name_node= node.children[0]
        attrib_name = usfm_bytes[attrib_name_node.start_byte:attrib_name_node.end_byte] \
            .decode('utf-8').strip()
        if attrib_name == "|":
            attrib_name = self.DEFAULT_ATTRIB_MAP[node.parent.type]
        if attrib_name == "src": # for \fig
            attrib_name = "file"

        attrib_val_cap = self.usfm_language.query("((attributeValue) @attrib-val)").captures(node)
        if len(attrib_val_cap) > 0:
            attrib_value = usfm_bytes[attrib_val_cap[0][0].start_byte:\
                attrib_val_cap[0][0].end_byte].decode('utf-8').strip()
        else:
            attrib_value = ""
        parent_json_obj[attrib_name] = attrib_value

    def node_2_usj_table(self, node, usfm_bytes, parent_json_obj):
        '''Handle table related components and convert to usj'''
        if node.type == "table":
            table_json_obj = {"type": "table", "content":[]}
            for child in node.children:
                self.node_2_usj(child, usfm_bytes, table_json_obj)
            parent_json_obj['content'].append(table_json_obj)
        elif node.type == "tr":
            row_json_obj = {"type": "table:row:tr", "content":[]}
            for child in node.children[1:]:
                self.node_2_usj(child, usfm_bytes, row_json_obj)
            parent_json_obj['content'].append(row_json_obj)
        elif node.type in self.TABLE_CELL_MARKERS:
            tag_node = node.children[0]
            style = usfm_bytes[tag_node.start_byte:tag_node.end_byte].decode('utf-8')\
            .replace("\\","").strip()
            cell_json_obj = {"type": f"table:cell:{style}", "content":[]}
            if "r" in style:
                cell_json_obj["align"] = "end"
            else:
                cell_json_obj["align"] = "start"
            for child in node.children[1:]:
                self.node_2_usj(child, usfm_bytes, cell_json_obj)
            parent_json_obj['content'].append(cell_json_obj)

    def node_2_usj_milestone(self, node, usfm_bytes, parent_json_obj):
        '''create ms node in USX'''
        ms_name_cap = self.usfm_language.query('''(
            [(milestoneTag)
             (milestoneStartTag)
             (milestoneEndTag)
             (zSpaceTag)
             ] @ms-name)''').captures(node)[0]
        style = usfm_bytes[ms_name_cap[0].start_byte:ms_name_cap[0].end_byte].decode('utf-8')\
        .replace("\\","").strip()
        ms_json_obj = {"type": f"ms:{style}", "content":[]}
        for child in node.children:
            if child.type.endswith("Attribute"):
                self.node_2_usj(child, usfm_bytes, ms_json_obj)
        parent_json_obj['content'].append(ms_json_obj)

    def node_2_usj_special(self, node, usfm_bytes, parent_json_obj):
        '''Build nodes for esb, cat, fig, optbreak in USX'''
        if node.type == "esb":
            sidebar_json_obj = {"type": "sidebar:esb", "content":[]}
            for child in node.children[1:-1]:
                self.node_2_usj(child, usfm_bytes, sidebar_json_obj)
            parent_json_obj['content'].append(sidebar_json_obj)
        elif node.type == "cat":
            cat_cap = self.usfm_language.query('((category) @category)').captures(node)[0]
            category = usfm_bytes[cat_cap[0].start_byte:cat_cap[0].end_byte].decode('utf-8').strip()
            parent_json_obj['category'] = category
        elif node.type == 'fig':
            fig_json_obj = {"type":"figure:fig", "content":[]}
            for child in node.children[1:-1]:
                self.node_2_usj(child, usfm_bytes, fig_json_obj)
            parent_json_obj['content'].append(fig_json_obj)
        elif node.type == 'b':
            b_json_obj = {"type": "optbreak:b", "content":[]}
            parent_json_obj['content'].append(b_json_obj)

    def node_2_usj_generic(self, node, usfm_bytes, parent_json_obj):
        '''build nodes for para style markers in USX'''
        tag_node = node.children[0]
        style = usfm_bytes[tag_node.start_byte:tag_node.end_byte].decode('utf-8')
        if style.startswith('\\'):
            style = style.replace('\\','').strip()
        else:
            style = node.type
        children_range_start = 1
        if len(node.children)>1 and node.children[1].type.startswith("numbered"):
            num_node = node.children[1]
            num = usfm_bytes[num_node.start_byte:num_node.end_byte].decode('utf-8')
            style += num
            children_range_start = 2
        para_json_obj = {"type": f"para:{style}", "content":[]}
        for child in node.children[children_range_start:]:
            if child.type in self.CHAR_STYLE_MARKERS+self.NESTED_CHAR_STYLE_MARKERS+\
            ["text", "footnote", "crossref", "verseText", "v", "b", "milestone", "zNameSpace"]:
            # only nest these types inside the upper para style node
                self.node_2_usj(child, usfm_bytes, para_json_obj)
            else:
                self.node_2_usj(child, usfm_bytes, parent_json_obj)
        parent_json_obj['content'].append(para_json_obj)

    def node_2_usj(self, node, usfm_bytes, parent_json_obj): # pylint: disable= too-many-branches
        '''check each node and based on the type convert to corresponding xml element'''
        # print("working with node: ", node, "\n")
        if node.type == "id":
            self.node_2_usj_id(node, usfm_bytes, parent_json_obj)
        elif node.type == "chapter":
            self.node_2_usj_chapter(node, usfm_bytes,parent_json_obj)
        elif node.type in ["cl", "cp", "cd", "vp"]:
            self.node_2_usj_generic(node, usfm_bytes, parent_json_obj)
        elif node.type in ["ca", "va"]:
            self.node_2_usj_ca_va(node, usfm_bytes, parent_json_obj)
        elif node.type == "v":
            self.node_2_usj_verse(node, usfm_bytes, parent_json_obj)
        elif node.type == "verseText":
            for child in node.children:
                self.node_2_usj(child, usfm_bytes, parent_json_obj)
        elif node.type in ['paragraph', 'pi', "ph"]:
            self.node_2_usj_para(node, usfm_bytes, parent_json_obj)
        elif node.type in self.NOTE_MARKERS:
            self.node_2_usj_notes(node, usfm_bytes, parent_json_obj)
        elif node.type in self.CHAR_STYLE_MARKERS+self.NESTED_CHAR_STYLE_MARKERS+["xt_standalone"]:
            self.node_2_usj_char(node, usfm_bytes, parent_json_obj)
        elif node.type.endswith("Attribute"):
            self.node_2_usj_attrib(node, usfm_bytes, parent_json_obj)
        elif node.type == 'text':
            text_val = usfm_bytes[node.start_byte:node.end_byte].decode('utf-8').strip()
            parent_json_obj['content'].append(text_val)
        elif node.type in ["table", "tr"]+ self.TABLE_CELL_MARKERS:
            self.node_2_usj_table(node, usfm_bytes, parent_json_obj)
        elif  node.type == "milestone":
            self.node_2_usj_milestone(node, usfm_bytes, parent_json_obj)
        elif node.type == "zNameSpace":
            self.node_2_usj_milestone(node, usfm_bytes, parent_json_obj)
        elif node.type in ["esb", "cat", "fig", "b"]:
            self.node_2_usj_special(node, usfm_bytes, parent_json_obj)
        elif (node.type in self.PARA_STYLE_MARKERS or
              node.type.replace("\\","").strip() in self.PARA_STYLE_MARKERS):
            self.node_2_usj_generic(node, usfm_bytes, parent_json_obj)
        elif node.type.strip() in ["","|"]:
            pass # skip white space nodes
        elif len(node.children)>0:
            for child in node.children:
                self.node_2_usj(child, usfm_bytes, parent_json_obj)
        # else:
        #     raise Exception("Encountered unknown element ", str(node))

File: node-usfm-parser/src/usjGenerator.js
//Logics for syntax-tree to dict(USJ) conversions
const Parser = require("tree-sitter");
const {Query} = Parser;

const { PARA_STYLE_MARKERS, NOTE_MARKERS, CHAR_STYLE_MARKERS, NESTED_CHAR_STYLE_MARKERS, DEFAULT_ATTRIB_MAP, TABLE_CELL_MARKERS, MISC_MARKERS } = require("./utils/markers");
class USJGenerator {


  constructor(treeSitterLanguageObj, usfmString, usjRootObj=null) {
    this.usfmLanguage = treeSitterLanguageObj;
    this.usfm = usfmString;
    this.jsonRootObj = usjRootObj || {
      type: "USJ",
      version: "0.3.0",
      content: [],
    };
  }

  findLastFromJson(jsonObj, typeValue) {
    let output = null;
    if (
      typeValue === jsonObj.type ||
      (jsonObj.marker && typeValue === jsonObj.marker)
    ) {
      output = jsonObj;
    }
    if (jsonObj.content) {
      jsonObj.content.forEach((child) => {
        if (typeof child === "string") {
          return;
        }
        const childOutput = this.findLastFromJson(child, typeValue);
        if (childOutput !== null) {
          output = childOutput;
        }
      });
    }
    return output;
  }

  nodeToUSJId(node, parentJsonObj) {
    const idCaptures = new Query(this.usfmLanguage, 
      "(id (bookcode) @book-code (description)? @desc)")
      .captures(node);
    let code = null;
    let desc = null;
    idCaptures.forEach((capture) => {
      if (capture.name === "book-code") {
        code = this.usfm.slice(capture.node.startIndex, capture.node.endIndex);
      } else if (capture.name === "desc") {
        desc = this.usfm.slice(capture.node.startIndex, capture.node.endIndex);
      }
    });
    const bookJsonObj = {
      type: "book",
      marker: "id",
      code: code,
      content: [],
    };
    if (desc && desc.trim() !== "") {
      bookJsonObj.content.push(desc.trim());
    }
    parentJsonObj.content.push(bookJsonObj);
  }

  // Similar conversion methods for other node types
  nodeToUSJC(node, parentJsonObj) {
    // Build c, the chapter milestone node in usj
    const chapCap = new Query(this.usfmLanguage,
        `(c (chapterNumber) @chap-num
                                             (ca (chapterNumber) @alt-num)?
                                             (cp (text) @pub-num)?)`,
      )
      .captures(node);
    const chapNum = this.usfm.slice(
      chapCap[0].node.startIndex,
      chapCap[0].node.endIndex,
    );
    let chapRef = null;
    this.jsonRootObj.content.forEach((child) => {
      if (child.type === "book") {
        chapRef = `${child.code} ${chapNum}`;
        return;
      }
    });

    const chapJsonObj = {
      type: "chapter",
      marker: "c",
      number: chapNum,
      sid: chapRef,
    };

    chapCap.forEach((tuple) => {
      if (tuple[1] === "alt-num") {
        chapJsonObj.altnumber = this.usfm
          .substring(tuple[0].startIndex, tuple[0].endIndex)
          .trim();
      }
      if (tuple[1] === "pub-num") {
        chapJsonObj.pubnumber = this.usfm
          .substring(tuple[0].startIndex, tuple[0].endIndex)
          .trim();
      }
    });

    parentJsonObj.content.push(chapJsonObj);

    node.children.forEach((child) => {
      if (["cl", "cd"].includes(child.type)) {
        this.nodeToUSJ(child, parentJsonObj);
      }
    });
  }

  nodeToUSJChapter(node, parentJsonObj) {
    // Build chapter node in USJ
    node.children.forEach((child) => {
      if (child.type === "c") {
        this.nodeToUSJC(child, parentJsonObj);
      } else {
        this.nodeToUSJ(child, parentJsonObj);
      }
    });
  }

  nodeToUSJVerse(node, parentJsonObj) {
    // Build verse node in USJ
    const verseNumCap = new Query(this.usfmLanguage,
        `
      (v
          (verseNumber) @vnum
          (va (verseNumber) @alt)?
          (vp (text) @vp)?
      )`,
      )
      .captures(node);

    const verseNum = this.usfm.substring(
      verseNumCap[0].node.startIndex,
      verseNumCap[0].node.endIndex,
    );

    const vJsonObj = {
      type: "verse",
      marker: "v",
      number: verseNum.trim(),
    };

    verseNumCap.forEach((capture) => {
      if (capture.name === "alt") {
        const altNum = this.usfm.slice(
          capture.node.startIndex,
          capture.node.endIndex,
        );
        vJsonObj.altnumber = altNum;
      } else if (capture.name === "vp") {
        const vpText = this.usfm.substring(
          capture.node.startIndex,
          capture.node.endIndex,
        );
        vJsonObj.pubnumber = vpText;
      }
    });

    const ref = `${this.findLastFromJson(this.jsonRootObj, "chapter").sid}:${verseNum}`;
    vJsonObj.sid = ref.trim();

    parentJsonObj.content.push(vJsonObj);
  }

  nodeToUSJCaVa(node, parentJsonObj) {
    // Build elements for independent ca and va away from c and v
    const style = node.type;
    const charJsonObj = {
      type: "char",
      marker: style,
    };

    const altNumMatch = new Query(this.usfmLanguage,
        `([
        (chapterNumber)
        (verseNumber)
    ] @alt-num)`,
      )
      .captures(node);

    const altNum = this.usfm
      .slice(altNumMatch[0].node.startIndex, altNumMatch[0].node.endIndex)
      .trim();

    charJsonObj.altnumber = altNum;
    parentJsonObj.content.push(charJsonObj);
  }

  nodeToUSJPara(node, parentJsonObj) {
    // Build paragraph nodes in USJ
    if (node.children[0].type.endsWith("Block")) {
      node.children[0].children.forEach((child) => {
        this.nodeToUSJPara(child, parentJsonObj);
      });
    } else if (node.type === "paragraph") {
      const paraTagCap = new Query(this.usfmLanguage,
        "(paragraph (_) @para-marker)")
        .captures(node)[0];
      const paraMarker = paraTagCap.node.type;

      if (paraMarker === "b") {
        this.nodeToUSJSpecial(paraTagCap, parentJsonObj);
      } else if (!paraMarker.endsWith("Block")) {
        const paraJsonObj = { type: "para", marker: paraMarker, content: [] };
        paraTagCap.node.children.forEach((child) => {
          this.nodeToUSJ(child, paraJsonObj);
        });
        parentJsonObj.content.push(paraJsonObj);
      }
    } else if (["pi", "ph"].includes(node.type)) {
      const paraMarker = this.usfm
        .substring(node.children[0].startIndex, node.children[0].endIndex)
        .replace("\\", "")
        .trim();
      const paraJsonObj = { type: "para", marker: paraMarker, content: [] };
      node.children.slice(1).forEach((child) => {
        this.nodeToUSJ(child, paraJsonObj);
      });
      parentJsonObj.content.push(paraJsonObj);
    }
  }

  nodeToUSJNotes(node, parentJsonObj) {
    // Build USJ nodes for footnotes and cross-references
    const tagNode = node.children[0];
    const callerNode = node.children[1];
    const style = this.usfm
      .substring(tagNode.startIndex, tagNode.endIndex)
      .replace("\\", "")
      .trim();
    const noteJsonObj = {
      type: "note",
      marker: style,
      content: [],
    };

    noteJsonObj.caller = this.usfm
      .substring(callerNode.startIndex, callerNode.endIndex)
      .trim();

    for (let i = 2; i < node.children.length - 1; i++) {
      this.nodeToUSJ(node.children[i], noteJsonObj);
    }

    parentJsonObj.content.push(noteJsonObj);
  }

  nodeToUSJChar(node, parentJsonObj) {
    // Build USJ nodes for character markups, both regular and nested
    const tagNode = node.children[0];
    let childrenRange = node.children.length;
    if (node.children[node.children.length - 1].type.startsWith("\\")) {
      childrenRange -= 1; // Exclude the last node if it starts with '\', treating it as a closing node
    }
    const style = this.usfm
      .substring(tagNode.startIndex, tagNode.endIndex)
      .replace("\\", "")
      .replace("+", "")
      .trim();
    const charJsonObj = {
      type: "char",
      marker: style,
      content: [],
    };

    // Assume a flag for closed markup, toggle this if your conditions and data structure require
    // charJsonObj.closed = node.children[node.children.length - 1].type.startsWith('\\');

    for (let i = 1; i < childrenRange; i++) {
      this.nodeToUSJ(node.children[i], charJsonObj);
    }

    parentJsonObj.content.push(charJsonObj);
  }

  nodeToUSJTable(node, parentJsonObj) {
    // Handle table related components and convert to USJ
    if (node.type === "table") {
      const tableJsonObj = { type: "table", content: [] };
      node.children.forEach((child) => {
        this.nodeToUSJ(child, tableJsonObj);
      });
      parentJsonObj.content.push(tableJsonObj);
    } else if (node.type === "tr") {
      const rowJsonObj = { type: "table:row", marker: "tr", content: [] };
      node.children.slice(1).forEach((child) => {
        this.nodeToUSJ(child, rowJsonObj);
      });
      parentJsonObj.content.push(rowJsonObj);
    } else if (TABLE_CELL_MARKERS.includes(node.type)) {
      const tagNode = node.children[0];
      const style = this.usfm
        .substring(tagNode.startIndex, tagNode.endIndex)
        .replace("\\", "")
        .trim();
      const cellJsonObj = {
        type: "table:cell",
        marker: style,
        content: [],
        align: style.includes("r") ? "end" : "start",
      };
      node.children.slice(1).forEach((child) => {
        this.nodeToUSJ(child, cellJsonObj);
      });
      parentJsonObj.content.push(cellJsonObj);
    }
  }

  nodeToUSJAttrib(node, parentJsonObj) {
    // Add attribute values to USJ elements
    const attribNameNode = node.children[0];
    let attribName = this.usfm
      .slice(attribNameNode.startIndex, attribNameNode.endIndex)
      .trim();

    // Handling special cases for attribute names
    if (attribName === "|") {
      attribName = DEFAULT_ATTRIB_MAP[node.parent.type];
    }
    if (attribName === "src") {
      // for \fig
      attribName = "file";
    }

    const attribValCap = new Query(this.usfmLanguage,
      "((attributeValue) @attrib-val)")
      .captures(node);

    let attribValue = "";
    if (attribValCap.length > 0) {
      attribValue = this.usfm
        .substring(
          attribValCap[0].node.startIndex,
          attribValCap[0].node.endIndex,
        )
        .trim();
    }

    parentJsonObj[attribName] = attribValue;
  }

  nodeToUSJMilestone(node, parentJsonObj) {
    // Create ms node in USJ

    const msNameCap = new Query(this.usfmLanguage,
        `(
        [(milestoneTag)
         (milestoneStartTag)
         (milestoneEndTag)
         (zSpaceTag)
         ] @ms-name)`,
      )
      .captures(node)[0];

    const style = this.usfm
      .slice(msNameCap.node.startIndex, msNameCap.node.endIndex)
      .replace("\\", "")
      .trim();
    const msJsonObj = { type: "ms", marker: style, content: [] };

    node.children.forEach((child) => {
      if (child.type.endsWith("Attribute")) {
        this.nodeToUSJ(child, msJsonObj);
      }
    });

    // Though normally milestones don't have contents, custom z-namespaces could have them
    if (!msJsonObj.content.length) {
      delete msJsonObj.content; // Remove empty content array if not used
    }

    parentJsonObj.content.push(msJsonObj);
  }

  nodeToUSJSpecial(node, parentJsonObj) {
    // Build nodes for esb, cat, fig, optbreak in USJ

    if (node.type === "esb") {
      const sidebarJsonObj = { type: "sidebar", marker: "esb", content: [] };
      node.children.slice(1, -1).forEach((child) => {
        this.nodeToUSJ(child, sidebarJsonObj);
      });
      parentJsonObj.content.push(sidebarJsonObj);
    } else if (node.type === "cat") {
      const catCap = new Query(this.usfmLanguage,
        "((category) @category)")
        .captures(node)[0];
      const category = this.usfm
        .substring(catCap[0].startIndex, catCap[0].endIndex)
        .trim();
      parentJsonObj.category = category;
    } else if (node.type === "fig") {
      const figJsonObj = { type: "figure", marker: "fig", content: [] };
      node.children.slice(1, -1).forEach((child) => {
        this.nodeToUSJ(child, figJsonObj);
      });
      parentJsonObj.content.push(figJsonObj);
    } else if (node.type === "b") {
      const bJsonObj = { type: "optbreak", marker: "b" };
      parentJsonObj.content.push(bJsonObj);
    } else if (node.type === "usfm") {
      const verJsonObj = { type: "para", marker: "usfm", content: [] };
      const version = this.usfm
        .substring(node.startIndex, node.endIndex)
        .replace("\\usfm", "")
        .trim();
      verJsonObj.content.push(version);
      parentJsonObj.content.push(verJsonObj);
    }
  }
  nodeToUSJGeneric(node, parentJsonObj) {
    // Build nodes for para style markers in USJ
    const tagNode = node.children[0];

    let style = this.usfm.substring(tagNode.startIndex, tagNode.endIndex);
    if (style.startsWith("\\")) {
      style = style.replace("\\", "").trim();
    } else {
      style = node.type;
    }

    // console.log(node.children.length, node.children[0].type, node.children[1].type)
    let childrenRangeStart = 1;
    if (
      node.children.length > 1 &&
      node.children[1].type.startsWith("numbered")
    ) {
      const numNode = node.children[1];
      const num = this.usfm.substring(numNode.startIndex, numNode.endIndex);
      style += num;
      childrenRangeStart = 2;
    }
    const paraJsonObj = { type: "para", marker: style, content: [] };
    parentJsonObj.content.push(paraJsonObj);

    for (let i = childrenRangeStart; i < node.children.length; i++) {
      const child = node.children[i];
      if (
        CHAR_STYLE_MARKERS.includes(child.type) ||
        NESTED_CHAR_STYLE_MARKERS.includes(child.type) ||
        [
          "text",
          "footnote",
          "crossref",
          "verseText",
          "v",
          "b",
          "milestone",
          "zNameSpace",
        ].includes(child.type)
      ) {
        // Only nest these types inside the upper para style node
        this.nodeToUSJ(child, paraJsonObj);
      } else {
        this.nodeToUSJ(child, parentJsonObj);
      }
    }
  }

  nodeToUSJ(node, parentJsonObj) {
    // Check each node and based on the type convert to corresponding XML element
    switch (node.type) {
      case "id":
        this.nodeToUSJId(node, parentJsonObj);
        break;
      case "chapter":
        this.nodeToUSJChapter(node, parentJsonObj);
        break;
      case "cl":
      case "cp":
      case "cd":
      case "vp":
        this.nodeToUSJGeneric(node, parentJsonObj);
        break;
      case "ca":
      case "va":
        this.nodeToUSJCaVa(node, parentJsonObj);
        break;
      case "v":
        this.nodeToUSJVerse(node, parentJsonObj);
        break;
      case "verseText":
        node.children.forEach((child) => this.nodeToUSJ(child, parentJsonObj));
        break;
      case "paragraph":
      case "pi":
      case "ph":
        this.nodeToUSJPara(node, parentJsonObj);
        break;
      case "text":
        const textVal = this.usfm
          .substring(node.startIndex, node.endIndex)
          .trim();
        if (textVal !== "") {
          parentJsonObj.content.push(textVal);
        }
        break;
      case "table":
      case "tr":
        this.nodeToUSJTable(node, parentJsonObj);
        break;
      case "milestone":
      case "zNameSpace":
        this.nodeToUSJMilestone(node, parentJsonObj);
        break;
      case "esb":
      case "cat":
      case "fig":
      case "usfm":
        this.nodeToUSJSpecial(node, parentJsonObj);
        break;
      default:
        if (
          CHAR_STYLE_MARKERS.includes(node.type) ||
          NESTED_CHAR_STYLE_MARKERS.includes(node.type) ||
          ["xt_standalone"].includes(node.type)
        ) {
          this.nodeToUSJChar(node, parentJsonObj);
        } else if (node.type.endsWith("Attribute")) {
          this.nodeToUSJAttrib(node, parentJsonObj);
        } else if (
          PARA_STYLE_MARKERS.includes(node.type) ||
          PARA_STYLE_MARKERS.includes(
            node.type.replace("\\", "").trim(),
          )
        ) {
          this.nodeToUSJGeneric(node, parentJsonObj);
        } else if (["", "|"].includes(node.type.trim())) {
          // Skip white space nodes
          break;
        } else if (node.children.length > 0) {
          node.children.forEach((child) =>
            this.nodeToUSJ(child, parentJsonObj),
          );
        }
        //  else {
        //
        //   console.error("Encountered unknown element ", node.type);

        // }
        break;
    }
  }
}

exports.USJGenerator = USJGenerator;
